cmd/libsnap-confine-private/mountinfo.c

```diff
@@ -126,42 +127,86 @@ static void show_buffers(const char *line, int offset,
 	fprintf(stderr, ">%s<\n", line);
 
 	fputc('>', stderr);
-	for (int i = 0; i < strlen(line); ++i) {
+	for (size_t i = 0; i < strlen(line); ++i) {
 		int c = entry->line_buf[i];
 		fputc(c == 0 ? '@' : c == 1 ? '#' : c, stderr);
 	}
 	fputc('<', stderr);
 	fputc('\n', stderr);
 
 	fputc('>', stderr);
-	for (int i = 0; i < strlen(line); ++i)
+	for (size_t i = 0; i < strlen(line); ++i)
 		fputc('=', stderr);
 	fputc('<', stderr);
 	fputc('\n', stderr);
 #endif				// MOUNTINFO_DEBUG
 }
 
+static bool is_octal_digit(char c)
+{
+	return c >= '0' && c <= '7';
+}
+
 static char *parse_next_string_field(sc_mountinfo_entry * entry,
-				     const char *line, int *offset)
+				     const char *line, size_t * offset)
 {
-	int offset_delta = 0;
-	char *field = &entry->line_buf[0] + *offset;
-	if (line[*offset] == ' ') {
-		// Special case for empty fields which cannot be parsed with %s.
-		*field = '\0';
-		*offset += 1;
-	} else {
-		int nscanned =
-		    sscanf(line + *offset, "%s%n", field, &offset_delta);
-		if (nscanned != 1)
-			return NULL;
-		*offset += offset_delta;
-		if (line[*offset] == ' ') {
-			*offset += 1;
+	const char *input = &line[*offset];
+	char *output = &entry->line_buf[*offset];
+	const char *s;		// Used for escape logic.
+	size_t input_idx = 0;
+	size_t output_idx = 0;
+
+	// Scan characters until we run out of memory to scan or we find a space.
+	// The kernel uses simple octal escape sequences for the following:
+	// space, tab, newline, backwards slash. Everything else is copied
+	// verbatim.
+	bool stop = false;
+	for (int c; c = input[input_idx], c != '\0' && !stop;
+	     input_idx++, output_idx++) {
+		switch (c) {
+		case ' ':
+			// Fields are space delimited. Represent the space
+			// as the end-of-string marker, skip over it, and stop parsing.
+			output[output_idx] = '\0';
+			stop = true;
+			break;
+		case '\\':
+			// Three *more* octal digits required for the escape sequence.
```

Comment:
I know the point of this is to unescape the space so we can continue parsing, but I wonder if it makes sense to leave the escape sequence in place for everything except the space, so we don't have to worry about weird stuff. The kernel isn't escaping too much, so maybe not a big deal, but unescaping the tab, newline and forward slash seems unneeded (ie, perhaps we could add unescaping those when we really wanted them).

On the fence if this is a blocker, but I consider it a hardening measure of sorts.

Reply:
Actually the point is not to unescape the space. The space is not a problem.

The space handling is now fixed in the loop, if we don't unescape anything the fix remains correct. I chose to implement the unescape logic for correctness and completeness of the logic only.

Reply:
If there is no benefit to us for unescaping, maybe just drop the code with a comment that the line might have escaped characters, but we currently don't care and are just passing them through?

Reply:
> If there is no benefit to us for unescaping

That said, I personally found the unescaping logic very nice :)